launchpad/src/launchpad.rs

```diff
@@ -1,9 +1,500 @@
+        let owner = self.blockchain().get_caller();
+
+        if sc_balance > 0 {
+            self.send()
+                .direct(&owner, &ticket_payment_token, 0, &sc_balance, &[]);
+        }
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[endpoint(forceClaimPeriodStart)]
+    fn force_claim_period_start(&self) -> SCResult<()> {
+        let total_winning_tickets = self.nr_winning_tickets().get();
+        self.total_confirmed_tickets().set(&total_winning_tickets);
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[endpoint(addTickets)]
+    fn add_tickets(
+        &self,
+        #[var_args] address_number_pairs: VarArgs<MultiArg2<Address, usize>>,
+    ) -> SCResult<()> {
+        self.require_no_ongoing_operation()?;
+        self.require_stage(LaunchStage::AddTickets)?;
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        let winner_selection_start_epoch = self.winner_selection_start_epoch().get();
+        require!(
+            current_epoch < winner_selection_start_epoch,
+            "Cannot add more tickets, winner selection has started"
+        );
+
+        for multi_arg in address_number_pairs.into_vec() {
+            let (buyer, nr_tickets) = multi_arg.into_tuple();
+
+            self.try_create_tickets(&buyer, nr_tickets)?;
+        }
+
+        Ok(())
+    }
+
+    // endpoints
+
+    #[endpoint(selectWinners)]
+    fn select_winners(&self) -> SCResult<OperationCompletionStatus> {
+        self.require_stage(LaunchStage::SelectWinners)?;
+
+        let last_ticket_position = self.shuffled_tickets().len();
+        let nr_winning_tickets = self.nr_winning_tickets().get();
+
+        // dummy, will be overwritten in the Load part,
+        // but the Rust compiler complains of possibly uninitialized variables otherwise
+        let mut rng = Random::from_hash(self.crypto(), H256::zero(), 0);
+        let mut ticket_position = 0;
+
+        let run_result = self.run_while_it_has_gas(|gas_op| match gas_op {
+            GasOp::Load(op) => match op {
+                OngoingOperationType::None => {
+                    rng = Random::from_seeds(
+                        self.crypto(),
+                        self.blockchain().get_prev_block_random_seed(),
+                        self.blockchain().get_block_random_seed(),
+                    );
+                    ticket_position = VEC_MAPPER_START_INDEX;
+
+                    LoopOp::Continue
+                }
+                OngoingOperationType::SelectWinners {
+                    seed,
+                    seed_index,
+                    ticket_position: ticket_pos,
+                } => {
+                    rng = Random::from_hash(self.crypto(), seed, seed_index);
+                    ticket_position = ticket_pos;
+
+                    LoopOp::Continue
+                }
+                _ => LoopOp::Break,
+            },
+            GasOp::Continue => {
```

Comment:
I think you should have only this operation - gasOpContinue, which calls Save if out of gas. GasOp Completed should be outside of the run_while_it_has_gas